### PR TITLE
Initial Auth Impl + More Caching

### PR DESCRIPTION
**Pull request by Geometrically**

### CODE REVIEW

File: src/auth/session.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        .headers()
+        .get("cf-ipcountry")
+        .and_then(|x| x.to_str().ok());
+    let city = req.headers().get("cf-ipcity").and_then(|x| x.to_str().ok());
+
+    let user_agent = req
+        .headers()
+        .get("user-agent")
+        .and_then(|x| x.to_str().ok())
+        .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
```
Comment (thesuzerain) [inline src/auth/session.rs:38]: Could we make this a separate err? If I were trying to log in and I forgot the user agent, invalid credentials being the error would be confusing and make me think I had the wrong password or mistyped my token or smthn.

If we dont anticipate this being used by the API ever directly then its fine

File: src/auth/validate.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        &redis,
+    )
+    .await?
+    .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
+
+    let user = User {
+        id: UserId::from(db_user.id),
+        github_id: db_user.github_id.map(|x| x as u64),
+        // discord_id: minos_user.discord_id,
+        // google_id: minos_user.google_id,
```
Comment (thesuzerain) [inline src/auth/validate.rs:35]: We can delete these now

File: src/auth/validate.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    executor: E,
+    redis: &deadpool_redis::Pool,
+) -> Result<User, AuthenticationError>
+where
+    E: sqlx::Executor<'a, Database = sqlx::Postgres> + Copy,
+{
+    let token: Option<&HeaderValue> = headers.get(AUTHORIZATION);
+
+    // Fetch DB user record and minos user from headers
+    let db_user = get_user_record_from_bearer_token(
```
Comment (thesuzerain) [inline src/auth/validate.rs:20]: Cookie implementation is still TODO, right? Using the bearer token as the primary authentication point is totally good but we should definitely have a way to get the mra_ bearer token from a cookie

File: src/database/models/session_item.rs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+    pub async fn get_many<'a, E, T: ToString>(
+        session_strings: &[T],
+        exec: E,
+        redis: &deadpool_redis::Pool,
+    ) -> Result<Vec<Session>, DatabaseError>
+    where
+        E: sqlx::Executor<'a, Database = sqlx::Postgres>,
+    {
+        use futures::TryStreamExt;
```
Comment (thesuzerain) [inline src/database/models/session_item.rs:132]: clippy error here, unused use futures::TryStreamExt;

File: src/database/models/session_item.rs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        let res = cmd("GET")
+            .arg(format!("{}:{}", SESSIONS_USERS_NAMESPACE, user_id.0))
+            .query_async::<_, Option<Vec<i64>>>(&mut redis)
+            .await?;
+
+        if let Some(res) = res {
+            return Ok(res.into_iter().map(|x| SessionId(x)).collect());
+        }
+
+        use futures::TryStreamExt;
```
Comment (thesuzerain) [inline src/database/models/session_item.rs:268]: clippy

File: src/models/users.rs [hunk 6]
```diff
@@ -48,11 +47,11 @@ pub struct User {
...
     pub payout_data: Option<UserPayoutData>,
     pub github_id: Option<u64>,
-    pub discord_id: Option<u64>,
-    pub google_id: Option<u128>,
-    pub microsoft_id: Option<u64>,
-    pub apple_id: Option<u64>,
-    pub gitlab_id: Option<u64>,
+    // pub discord_id: Option<u64>,
+    // pub google_id: Option<u128>,
+    // pub microsoft_id: Option<u64>,
```
Comment (thesuzerain) [inline src/models/users.rs:52]: See above

File: src/models/users.rs [hunk 7]
```diff
@@ -147,11 +145,11 @@ impl From<DBUser> for User {
...
             payout_data: None,
             github_id: None,
-            discord_id: None,
-            google_id: None,
-            microsoft_id: None,
-            apple_id: None,
-            gitlab_id: None,
+            // discord_id: None,
+            // google_id: None,
+            // microsoft_id: None,
```
Comment (thesuzerain) [inline src/models/users.rs:150]: See above

**Review by thesuzerain — APPROVED** [review]

File: src/auth/session.rs (hunk 1 above)
Comment (Geometrically) [reply src/auth/session.rs:38]: it won't be in the API, so no need

File: src/auth/validate.rs (hunk 3 above)
Comment (Geometrically) [reply src/auth/validate.rs:20]: we won't support cookie auth due to csrf, the frontend supports this auth method

File: src/auth/validate.rs (hunk 2 above)
Comment (Geometrically) [reply src/auth/validate.rs:35]: i think I'll keep them in the future, for stuff like oauth linking

**Pull request merged**